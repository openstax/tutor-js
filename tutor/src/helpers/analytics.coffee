_ = require 'underscore'

Routes = require '../routes'

{CurrentUserStore} = require '../flux/current-user'

DestinationHelper  = require '../helpers/routes-and-destinations'

# generate custom event data for routes
Events =
  viewTaskStep: ({courseId}) ->
    # track all work done on a course
    Analytics.sendEvent 'Course', 'Work', label: courseId

  viewStudentDashboard: ({courseId}) ->
    # compare activity between courses
    Analytics.sendEvent 'Student', 'Dashboard', label: courseId


# Translators convert a url like '/foo/bar/123/baz/1' into a simplified one like just '/foo/bar'
Translators =

  dashboard:            ({courseId}) -> "/student/choose-course/#{courseId}"
  viewPractice:         ({courseId}) -> "/student/practice/#{courseId}"
  viewPerformanceForecast:  ({courseId}) -> "/student/performance-forecast/#{courseId}"
  viewStudentDashboard: ({courseId}) -> "/student/dashboard/#{courseId}"

  calendarByDate:        ({courseId}) -> "/teacher/calendar/#{courseId}"
  viewTeacherPerformanceForecast:      ({courseId}) -> "/teacher/performance-forecast/#{courseId}"
  viewScores:            ({courseId}) -> "/teacher/student-scores/#{courseId}"
  courseSettings:        ({courseId}) -> "/teacher/roster/#{courseId}"
  editReading:           ({courseId}) -> "/teacher/assignment/edit/reading/#{courseId}"
  editHomework:          ({courseId}) -> "/teacher/assignment/edit/homework/#{courseId}"
  editExternal:          ({courseId}) -> "/teacher/assignment/edit/external/#{courseId}"
  createReading:         ({courseId}) -> "/teacher/assignment/create/reading/#{courseId}"
  createHomework:        ({courseId}) -> "/teacher/assignment/create/homework/#{courseId}"
  createExternal:        ({courseId}) -> "/teacher/assignment/create/external/#{courseId}"
  calendarViewPlanStats: ({courseId}) -> "/teacher/metrics/quick/#{courseId}"
  reviewTask:            ({courseId}) -> "/teacher/metrics/review/#{courseId}"

  viewReferenceBook:        ({courseId}) -> "/reference-view/#{courseId}"
  viewReferenceBookSection: ({courseId}) -> "/reference-view/#{courseId}"
  viewReferenceBookPage:    ({courseId}) -> "/reference-view/#{courseId}"

  # Task steps are viewed by both teacher and student with no difference in params
  viewTaskStep:         ({courseId}) ->
    role = CurrentUserStore.getCourseRole(courseId, true)
    "/#{role}/task-step/#{courseId}"


GA = undefined

Analytics =

  setTracker: (tracker) -> GA = tracker

  sendPageView: (url) ->
    GA?('send', 'pageview', url)

  onNavigation: (path) ->
    return unless GA

    route = Routes.pathToEntry(path)

    return @sendPageView("/not-found/#{change.path}") unless route

    path = Translators[route.entry.name]?( route.match.params ) or route.match.pathname

    # if we're also going to send custom events then we set the page
    if Events[route.entry.name]
      GA('set', 'page', path)
      Events[route.entry.name]( route.match.params )
      @sendPageView() # url's not needed since it was set before events
    else
      @sendPageView(route.match.pathname)

  sendEvent: (category, action, attrs) ->
    return unless GA
    GA('send', 'event', category, action, attrs.label, attrs.value)

module.exports = Analytics
