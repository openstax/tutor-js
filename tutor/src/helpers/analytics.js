import { partial, get } from 'lodash';

import Router from '../helpers/router';
import Courses from '../models/courses-map';

// generate custom event data for routes
const Events = {
  viewTaskStep({ courseId }) {
    // track all work done on a course
    return Analytics.sendEvent('Course', 'Work', { label: courseId });
  },

  viewStudentDashboard({ courseId }) {
    // compare activity between courses
    return Analytics.sendEvent('Student', 'Dashboard', { label: courseId });
  },
};

// a bit shorter helper methods
const isTeacher = courseId => get(Courses.get(courseId), 'currentRole.isTeacher', false);
const getRole = function(courseId) {
  if (Courses.get(courseId).is_teacher) { return 'teacher'; } else { return 'student'; }
};

const assignmentTypeTranslator = function(assignmentType, { courseId, id }) {
  const newOrEdit = id === 'new' ? 'create' : 'edit';
  return `/teacher/assignment/${newOrEdit}/${assignmentType}/${courseId}`;
};

function viewReferenceBook({ courseId, pageId }) {
  const url = `/reference-view/${courseId}`;
  if (!pageId) {
    return url;
  }
  const course = Courses.get(courseId);
  const page = course.referenceBook.pages.byId.get(pageId);
  return page ? `${url}/section/${page.chapter_section.key}` : url;
}

// Translators convert a url like '/foo/bar/123/baz/1' into a simplified one like just '/foo/bar'
const Translators = {
  dashboard({ courseId }) {
    if (isTeacher(courseId)) { return `/teacher/calendar/${courseId}`; } else { return `/student/dashboard/${courseId}`; }
  },
  practiceTopics({ courseId }) { return `/student/practice/${courseId}`; },
  viewPerformanceGuide({ courseId }) {
    return `/${getRole(courseId)}/performance-forecast/${courseId}`;
  },
  calendarByDate({ courseId }) { return `/teacher/calendar/${courseId}`; },
  viewGradebook({ courseId }) { return `/teacher/student-scores/${courseId}`; },
  courseSettings({ courseId }) { return `/teacher/roster/${courseId}`; },
  editAssignment: partial(assignmentTypeTranslator, 'assignment'),
  calendarViewPlanStats({ courseId }) { return `/teacher/metrics/quick/${courseId}`; },
  reviewTask({ courseId }) { return `/teacher/metrics/review/${courseId}`; },
  viewReferenceBookSection: viewReferenceBook,
  viewReferenceBookPage: viewReferenceBook,
  viewReferenceBook,

  // Task steps are viewed by both teacher and student with no difference in params
  viewTaskStep({ courseId }) {
    const course = Courses.get(courseId);
    const role = course ? course.primaryRole.type : 'unknown';
    return `/${role}/task-step/${courseId}`;
  },
};

const TAG_MANAGER_ID = 'GTM-W6N7PB';

let GA = undefined;
let trackerNames = undefined;

var Analytics = {
  // copied from BIT
  initTagManager(w, d, s, l, i) {
    // Disable ESLint rules since we're copying Google's script
    /* eslint-disable one-var, prefer-const, prefer-template */
    w[l] = w[l] || [];
    w[l].push({
      'gtm.start': new Date().getTime(),
      event: 'gtm.js',
    });
    let f = d.getElementsByTagName(s)[0], j = d.createElement(s), dl = l !== 'dataLayer' ? '&l=' + l : '';
    // Breaks in tests because there are no scripts
    if (f) {
      j.async = true;
      j.src = '//www.googletagmanager.com/gtm.js?id=' + i + dl;
      f.parentNode.insertBefore(j, f);
    }
  },
  
  // Set the Command Queue (ga) and cache the (possibly multiple) tracker names
  setGa(ga) {
    trackerNames = undefined;
    GA = ga;
    if (GA) {
      GA(function() {
        trackerNames = GA.getAll().map(function(tracker) { return tracker.get('name'); });
      });
      this.initTagManager(window, document, 'script', 'dataLayer', TAG_MANAGER_ID);
    }

    return GA;
  },

  // Calls all trackers with the command and arguments given here
  ga(command, ...params) {
    if (!GA || !trackerNames) { return; }
    trackerNames.forEach((trackerName) => {
      this.realGa(trackerName + '.' + command, ...params);
    });
  },

  // Exists largely for spec mocking
  realGa(...args) {
    GA(...args);
  },

  sendPageView(url) {
    this.ga('send', 'pageview', url);
  },

  recordCourseDimension(courseId) {
    const course = Courses.get(courseId);
    if (course) {
      this.ga('set', 'dimension1', course.appearance_code);
    }
  },

  onNavigation(path) {
    if (!GA) { return; }
    const route = Router.currentMatch(path);

    if (!route) {
      this.sendPageView(`/not-found${path}`);
      return;
    }
    const { courseId } = route.params;
    if (courseId) { this.recordCourseDimension(courseId); }
    const translatedPath = Translators[route.entry.name] ?
      Translators[route.entry.name]( route.params ) : route.pathname;

    this.ga('set', 'page', translatedPath);
    // if we're also going to send custom events then we set the page
    if (Events[route.entry.name]) {
      Events[route.entry.name]( route.params );
    }

    this.sendPageView(); // url's not needed since it was set before events
  },

  sendEvent(category, action, attrs = {}) {
    this.ga('send', 'event', category, action, attrs.label, attrs.value);
  },
};

export default Analytics;
