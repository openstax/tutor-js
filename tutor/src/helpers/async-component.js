import React from 'react';
import AsyncErrorBoundary from '../components/error-monitoring/async-load-error';
import LoadingAnimation from '../../../shared/src/components/loading-animation'

export function asyncComponent(loader, name = '') {

    const Component = React.lazy(loader);
    const loading = <LoadingAnimation message={`Loading ${name}â€¦`} />;

    const Loader = (props) => (
        <AsyncErrorBoundary>
            <React.Suspense fallback={loading}>
                <Component {...props} />
            </React.Suspense>
        </AsyncErrorBoundary>
    );

    return (props) => <Loader {...props} />;
}


export function loadAsync(resolve, name) {
    // return a function so the router will only evaluate it when it's needed
    // in the future we may insert role dependant logic here
    return () => asyncComponent(resolve, name);
}
