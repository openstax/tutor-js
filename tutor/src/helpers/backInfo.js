import { navHistory } from '../models'


const backInfo = () => {
    // Gets route to last path that was not the same as the current one
    // See TransitionStore for more detail.
    const historyInfo = navHistory.latest

    const text = historyInfo != null && !!historyInfo.name ? `Back to ${historyInfo.name}` : 'My Courses';
    const to =  historyInfo != null && !!historyInfo.path ? historyInfo.path : 'myCourses';

    return {
        text,
        to,
    };
};


export default backInfo;
