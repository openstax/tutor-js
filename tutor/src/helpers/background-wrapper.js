import { styled } from 'vendor';
import { navbars } from '../theme.js';

const BackgroundWrapper = styled.div`
  background: #fff;
  min-height: calc(100vh - ${navbars.top.height} - ${navbars.bottom.height});
  position: relative;
  overflow: hidden;
  padding: 0 2.4rem;
`;

const ContentWrapper = styled.div`
  max-width: 1200px;
  margin: 0 auto;
`;

export { BackgroundWrapper, ContentWrapper };
