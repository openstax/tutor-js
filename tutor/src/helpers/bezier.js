function EASE_IN_OUT(t) {
    if (t < .5) { return 4 * t * t * t; } else { return ((t - 1) * ((2 * t) - 2) * ((2 * t) - 2)) + 1; }
}

function CALC(start, end, elapsed, duration) {
    if (elapsed > duration) { return end; }
    return start + ((end - start) * EASE_IN_OUT(elapsed / duration));
}

export default function bezier({
    range, duration, onStep, onComplete,
}) {
    const startTime = Date.now();
    const requestAnimationFrame = window.requestAnimationFrame || window.setTimeout;

    const step = () => {
        const elapsed = Date.now() - startTime;
        onStep(CALC(range[0], range[1], elapsed, duration));
        if (elapsed < duration) {
            requestAnimationFrame(step);
        } else if (onComplete) {
            onComplete();
        }
    };

    step();
}
