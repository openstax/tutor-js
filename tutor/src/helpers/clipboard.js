import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
    isSupported: _.constant(!!SUPPORTED),

    copy() {
        if (!SUPPORTED) { return; }

        try {
            document.execCommand('copy');
        } catch (e) {
            // eslint-disable-next-line no-console
            console.warn('clipboard copy failed', e);
        }
    },
};
