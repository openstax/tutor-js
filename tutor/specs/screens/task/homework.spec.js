import UX from '../../../src/screens/task/ux';
import Homework from '../../../src/screens/task/homework';
import { TestRouter, Factory, C, FakeWindow, ld, TimeMock } from '../../helpers';

describe('Reading Tasks Screen', () => {
  let props, history;
  TimeMock.setTo('2017-10-14T12:00:00.000Z');

  beforeEach(() => {
    let task = Factory.studentTask({
      type: 'homework', stepCount: 5,
    });
    task.tasksMap = { course: Factory.course() }
    history = new TestRouter({
      push: (url) => {
        props.ux.goToStep(Number(ld.last(url.split('/'))) - 1, false);
      },
    }).history;
    props = {
      windowImpl: new FakeWindow(),
      ux: new UX({ task, course: Factory.course(), history }),
    };
  });

  it('matches snapshot', () => {
    expect(<C><Homework {...props} /></C>).toMatchSnapshot();
  });

  it('renders value props', () => {
    const h = mount(<C><Homework {...props} /></C>);
    expect(props.ux.currentStep.type).toEqual('task-instructions');
    props.ux.goForward();
    expect(props.ux.currentStep.type).toEqual('two-step-intro');
    expect(h).toHaveRendered('TwoStepValueProp');
    h.unmount();
  });

  it('renders task with placeholders', () => {
    props.ux.task.steps.forEach(s => s.type = 'placeholder');
    const h = mount(<C><Homework {...props} /></C>);
    expect(h).toHaveRendered('Instructions');
    props.ux.goForward();
    expect(h).toHaveRendered('IndividualReview');
    props.ux.isLocked = false;
    props.ux.goForward();
    h.render();
    expect(h).toHaveRendered('LoadingCard');
    props.ux.currentStep.api.requestCounts.read = 1;
    expect(h).toHaveRendered('PlaceHolderTaskStep');
    h.unmount();
  });
});
