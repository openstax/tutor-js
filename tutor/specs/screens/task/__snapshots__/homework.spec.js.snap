// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reading Tasks Screen After close date matches snapshot 1`] = `
<C>
  <HomeworkTask
    ux={
      TaskUX {
        "_stepId": "instructions",
        "_task": Object {
          "allow_auto_graded_multiple_attempts": false,
          "closes_at": "2017-10-11T12:00:00.000Z",
          "completion_weight": 0,
          "correctness_weight": 1,
          "description": "empowering",
          "due_at": "2017-10-04T12:00:00.000Z",
          "id": 5,
          "is_deleted": false,
          "is_provisional_score": true,
          "late_work_penalty_applied": "not_accepted",
          "late_work_penalty_per_period": 1,
          "published_late_work_point_penalty": 0,
          "published_points": 7,
          "spy": Object {},
          "steps": Array [
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 22,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 22,
              "is_completed": true,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "magnetic: Fusion",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "22@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 23,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 23,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "sexy: Substructure of the Nucleus",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "23@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 24,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 24,
              "is_completed": true,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "value-added: Fusion",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "24@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 25,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 25,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "leading-edge: Eukaryotic Post-transcriptional Gene Regulation",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "25@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 26,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 26,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "killer: Newton's First Law of Motion: Inertia",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "26@1",
            },
          ],
          "students": Array [],
          "title": "bi-directional",
          "type": "homework",
        },
        "course": Object {
          "appearance_code": "college_physics",
          "book_pdf_url": "https://archive.cnx.org/exports/185cbf87-c72e-48f5-b51e-f14f21b5eabd.pdf",
          "cloned_from_id": null,
          "code": "OX-1",
          "current_role_id": 0,
          "default_due_time": "07:00",
          "default_open_time": "00:32",
          "does_cost": false,
          "ecosystem_book_uuid": "185cbf87-c72e-48f5-b51e-f14f21b5eabd",
          "ecosystem_id": 12,
          "ends_at": "2017-11-14T13:00:00.000Z",
          "homework_progress_weight": 0.1,
          "homework_score_weight": 0.4,
          "homework_weight": 0.5,
          "id": 12,
          "is_access_switchable": true,
          "is_active": true,
          "is_college": true,
          "is_concept_coach": false,
          "is_lms_enabled": false,
          "is_lms_enabling_allowed": false,
          "is_preview": false,
          "just_created": false,
          "name": "College Physics",
          "offering_id": 1,
          "periods": Array [
            Object {
              "enrollment_code": "78629",
              "enrollment_url": "http://localhost:3001/enroll/78629",
              "id": 34,
              "is_archived": false,
              "name": "1st",
              "num_enrolled_students": 66,
            },
            Object {
              "enrollment_code": "78629",
              "enrollment_url": "http://localhost:3001/enroll/78629",
              "id": 35,
              "is_archived": false,
              "name": "2nd",
              "num_enrolled_students": 33,
            },
            Object {
              "enrollment_code": "78629",
              "enrollment_url": "http://localhost:3001/enroll/78629",
              "id": 36,
              "is_archived": false,
              "name": "3rd",
              "num_enrolled_students": 17,
            },
          ],
          "reading_progress_weight": 0.2,
          "reading_score_weight": 0.2,
          "reading_weight": 0.5,
          "roles": Array [
            Object {
              "id": 65,
              "joined_at": "2017-07-21T12:00:00.000Z",
              "period_id": 0,
              "research_identifier": "",
              "type": "student",
            },
          ],
          "salesforce_book_name": "College Physics",
          "should_reuse_preview": false,
          "starts_at": "2017-07-14T12:00:00.000Z",
          "students": Array [
            Object {
              "first_name": "Okey",
              "id": 12,
              "is_active": true,
              "is_comped": false,
              "is_paid": false,
              "is_refund_allowed": false,
              "is_refund_pending": false,
              "last_name": "DuBuque",
              "name": "",
              "payment_code": undefined,
              "payment_due_at": "2017-10-17T12:00:00.000Z",
              "period_id": 0,
              "prompt_student_to_pay": false,
              "role_id": 65,
              "student_identifier": "",
              "uuid": "",
            },
          ],
          "teacher_profiles": Array [
            Object {
              "id": 1,
              "name": "Review Teacher 1",
            },
            Object {
              "id": 2,
              "name": "Review Teacher 2",
            },
          ],
          "teachers": Array [],
          "term": "spring",
          "timezone": "US/Central",
          "uses_pre_wrm_scores": false,
          "uuid": "b35df724-3e75-47e1-88bd-76790c0657fc",
          "webview_url": "https://qa.cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd",
          "year": 2017,
        },
        "hideTaskProgressTable": true,
        "hideToolbar": false,
        "history": Object {
          "createHref": [MockFunction],
          "listen": [MockFunction],
          "location": Object {
            "pathname": "/",
          },
          "push": [Function],
          "replace": [MockFunction],
        },
        "isLocked": false,
        "viewedInfoSteps": Array [],
        "window": [Window],
      }
    }
    windowImpl={
      FakeWindow {
        "addEventListener": [MockFunction],
        "clearInterval": [MockFunction],
        "document": Object {
          "body": <body />,
          "hidden": false,
        },
        "getSelection": [MockFunction],
        "history": Object {
          "pushState": [MockFunction],
        },
        "innerHeight": 1024,
        "innerWidth": 768,
        "localStorage": Object {
          "getItem": [MockFunction],
          "setItem": [MockFunction],
        },
        "location": Object {
          "href": "http://localhost:3001/dashboard",
          "pathname": "/",
          "reload": [MockFunction],
          "search": "",
        },
        "open": [MockFunction],
        "pageXOffset": 0,
        "pageYOffset": 0,
        "querySelector": [MockFunction],
        "removeEventListener": [MockFunction],
        "requestAnimationFrame": [MockFunction],
        "screen": Object {
          "height": 1024,
          "width": 768,
        },
        "scroll": [MockFunction],
        "setInterval": [MockFunction],
      }
    }
  />
</C>
`;

exports[`Reading Tasks Screen Before close date matches snapshot 1`] = `
<C>
  <HomeworkTask
    ux={
      TaskUX {
        "_stepId": "instructions",
        "_task": Object {
          "allow_auto_graded_multiple_attempts": false,
          "closes_at": "2017-10-21T12:00:00.000Z",
          "completion_weight": 0,
          "correctness_weight": 1,
          "description": "background",
          "due_at": "2017-10-14T12:00:00.000Z",
          "id": 2,
          "is_deleted": false,
          "is_provisional_score": true,
          "late_work_penalty_applied": "assignment",
          "late_work_penalty_per_period": 0,
          "published_late_work_point_penalty": 0,
          "published_points": 9,
          "spy": Object {},
          "steps": Array [
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 7,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 7,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "best-of-breed: Simple Machines",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "7@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 8,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 8,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "proactive: Breathing",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "8@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 9,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 9,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "global: Eukaryotic Post-transcriptional Gene Regulation",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "9@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 10,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 10,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "front-end: Centripetal Acceleration",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "10@1",
            },
            Object {
              "answer_id": undefined,
              "attempt_number": 0,
              "attempts_remaining": 2,
              "available_points": 1,
              "can_be_updated": true,
              "correct_answer_feedback_html": "",
              "correct_answer_id": 0,
              "dropped_method": "",
              "exercise_id": 11,
              "external_url": null,
              "feedback_html": "",
              "formats": Array [
                "multiple-choice",
                "free-response",
              ],
              "free_response": "",
              "group": "",
              "id": 11,
              "is_completed": false,
              "is_feedback_available": false,
              "labels": undefined,
              "last_completed_at": "1970-01-01T00:00:00.000Z",
              "preview": "e-business: Newton's First Law of Motion: Inertia",
              "published_comments": "",
              "published_late_work_point_penalty": null,
              "published_points": null,
              "published_points_without_lateness": null,
              "response_validation": Object {},
              "solution": Object {},
              "spy": Object {},
              "tasked_id": 0,
              "type": "exercise",
              "uid": "11@1",
            },
          ],
          "students": Array [],
          "title": "foreground",
          "type": "homework",
        },
        "course": Object {
          "appearance_code": "college_physics",
          "book_pdf_url": "https://archive.cnx.org/exports/185cbf87-c72e-48f5-b51e-f14f21b5eabd.pdf",
          "cloned_from_id": null,
          "code": "OX-1",
          "current_role_id": 0,
          "default_due_time": "07:00",
          "default_open_time": "00:32",
          "does_cost": false,
          "ecosystem_book_uuid": "6c322e32-9fb0-4c4d-a1d7-20c95c5c7af2",
          "ecosystem_id": 3,
          "ends_at": "2017-11-14T13:00:00.000Z",
          "homework_progress_weight": 0.1,
          "homework_score_weight": 0.4,
          "homework_weight": 0.5,
          "id": 3,
          "is_access_switchable": true,
          "is_active": true,
          "is_college": true,
          "is_concept_coach": false,
          "is_lms_enabled": false,
          "is_lms_enabling_allowed": false,
          "is_preview": false,
          "just_created": false,
          "name": "College Physics",
          "offering_id": 1,
          "periods": Array [
            Object {
              "enrollment_code": "78629",
              "enrollment_url": "http://localhost:3001/enroll/78629",
              "id": 7,
              "is_archived": false,
              "name": "1st",
              "num_enrolled_students": 47,
            },
            Object {
              "enrollment_code": "78629",
              "enrollment_url": "http://localhost:3001/enroll/78629",
              "id": 8,
              "is_archived": false,
              "name": "2nd",
              "num_enrolled_students": 49,
            },
            Object {
              "enrollment_code": "78629",
              "enrollment_url": "http://localhost:3001/enroll/78629",
              "id": 9,
              "is_archived": false,
              "name": "3rd",
              "num_enrolled_students": 37,
            },
          ],
          "reading_progress_weight": 0.2,
          "reading_score_weight": 0.2,
          "reading_weight": 0.5,
          "roles": Array [
            Object {
              "id": 28,
              "joined_at": "2017-07-21T12:00:00.000Z",
              "period_id": 0,
              "research_identifier": "",
              "type": "student",
            },
          ],
          "salesforce_book_name": "College Physics",
          "should_reuse_preview": false,
          "starts_at": "2017-07-14T12:00:00.000Z",
          "students": Array [
            Object {
              "first_name": "Edwardo",
              "id": 3,
              "is_active": true,
              "is_comped": false,
              "is_paid": false,
              "is_refund_allowed": false,
              "is_refund_pending": false,
              "last_name": "Raynor",
              "name": "",
              "payment_code": undefined,
              "payment_due_at": "2017-10-17T12:00:00.000Z",
              "period_id": 0,
              "prompt_student_to_pay": false,
              "role_id": 28,
              "student_identifier": "",
              "uuid": "",
            },
          ],
          "teacher_profiles": Array [
            Object {
              "id": 1,
              "name": "Review Teacher 1",
            },
            Object {
              "id": 2,
              "name": "Review Teacher 2",
            },
          ],
          "teachers": Array [],
          "term": "spring",
          "timezone": "US/Central",
          "uses_pre_wrm_scores": false,
          "uuid": "3d9ee211-72b6-416d-8617-37c57ecd0a87",
          "webview_url": "https://qa.cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd",
          "year": 2017,
        },
        "hideTaskProgressTable": true,
        "hideToolbar": false,
        "history": Object {
          "createHref": [MockFunction],
          "listen": [MockFunction],
          "location": Object {
            "pathname": "/",
          },
          "push": [Function],
          "replace": [MockFunction],
        },
        "isLocked": false,
        "viewedInfoSteps": Array [],
        "window": [Window],
      }
    }
    windowImpl={
      FakeWindow {
        "addEventListener": [MockFunction],
        "clearInterval": [MockFunction],
        "document": Object {
          "body": <body />,
          "hidden": false,
        },
        "getSelection": [MockFunction],
        "history": Object {
          "pushState": [MockFunction],
        },
        "innerHeight": 1024,
        "innerWidth": 768,
        "localStorage": Object {
          "getItem": [MockFunction],
          "setItem": [MockFunction],
        },
        "location": Object {
          "href": "http://localhost:3001/dashboard",
          "pathname": "/",
          "reload": [MockFunction],
          "search": "",
        },
        "open": [MockFunction],
        "pageXOffset": 0,
        "pageYOffset": 0,
        "querySelector": [MockFunction],
        "removeEventListener": [MockFunction],
        "requestAnimationFrame": [MockFunction],
        "screen": Object {
          "height": 1024,
          "width": 768,
        },
        "scroll": [MockFunction],
        "setInterval": [MockFunction],
      }
    }
  />
</C>
`;
